File: suites/05_ml_primitives/comparisons/autodiff.py
def autodiff_trace(steps: int) -> int:
    value, grad = 7, 1
    for _ in range(steps):
        grad += (value * 3) % 17
        value += grad
    return value + grad
if __name__ == "__main__":
    print(autodiff_trace(5000000))

File: suites/05_ml_primitives/comparisons/autodiff.cpp
#include <cstdio>
static long long autodiff_trace(long long steps) {
    long long value = 7, grad = 1;
    for (long long s = 0; s < steps; ++s) {
        grad += (value * 3) % 17;
        value += grad;
    }
    return value + grad;
}
int main() { printf("%lld\n", autodiff_trace(5000000)); return 0; }
